pulp_rpm/app/downloaders.py

```diff
@@ -94,3 +104,87 @@ async def _run(self, extra_data=None):
             self.session.close()
         return to_return
+
+
+class UlnDownloader(RpmDownloader):
+    """
+    Custom Downloader for ULN repositories.
+
+    Args:
+        uln_user (str): Username for authentification in ULN network
+        uln_pwd (str): password for authentification in ULN network
+        uln_server_base_url (str): ULN server url. If no url is given, use default oracle
+            ULN server url `https://linux-update.oracle.com/`
+
+    Raises:
+        NoUlnCredentials: If no or not valid ULN credentials are given, this Error will be displayed
+    """
+
+    def __init__(self, *args, uln_user=None, uln_pwd=None, uln_server_base_url=None, **kwargs):
+        """
+        Initialize the downloader for ULN repositories.
+
+        If no server URL is given, use the server url from oracle linux.
+        """
+        self.uln_user = uln_user
+        self.uln_pwd = uln_pwd
+        if uln_server_base_url:
+            self.uln_server_base_url = uln_server_base_url
+        else:  # if no uln_server_base_url is given, use oracle linux url
+            self.uln_server_base_url = "https://linux-update.oracle.com/"
+        self.headers = None
+        self.session_key = None
+
+        super().__init__(*args, **kwargs)
+
+    async def _run(self, extra_data=None):
+        """
+        Download, validate, and compute digests on the `url`. This is a coroutine.
+
+        Once per session the coroutine logs into the ULN account using the
+        uln_username and uln_pwd. Ther returned key is used for authentification
+        for all other downloads.
+
+        This method provides the same return object type and documented in
+        :meth:`~pulpcore.plugin.download.BaseDownloader._run`.
+        """
+        if self.sles_auth_token:
+            auth_param = f"?{self.sles_auth_token}"
+            url = urljoin(self.url, auth_param)
+        else:
+            url = self.url
+
+        parsed = urlparse(self.url)
+        if parsed.scheme == "uln":
+            if not (self.uln_user and self.uln_pwd):
+                raise NoUlnCredentials()
+            # get ULN Session-key
+            SERVER_URL = os.path.join(self.uln_server_base_url, "rpc/api")
+
+            # if proxy, do some thing... TO DO
+
+            client = ServerProxy(SERVER_URL)
+            if not self.session_key:
+                self.session_key = await client["auth.login"](self.uln_user, self.uln_pwd)
+                if len(self.session_key) != 43:
+                    raise NoUlnCredentials()
+                self.headers = {"X-ULN-API-User-Key": self.session_key}
+                await client.close()
+            # build request url from input uri
+            channelLabel = parsed.netloc
+            path = parsed.path.lstrip("/")
+            url = os.path.join(self.uln_server_base_url, "XMLRPC/GET-REQ", channelLabel, path)
+            log.info(f"download_url = {url}\n with key = {self.session_key}")
```

Comment:
The f-string was introduced in Python 3.6 - does this code have to run on prior Python versions?
If this output is not mandatory at all, it can be left out.

Reply:
You are right. I left it out for now.